src/subscription-hoc.tsx

```diff
@@ -0,0 +1,116 @@
+import * as React from 'react';
+import { DocumentNode } from 'graphql';
+const hoistNonReactStatics = require('hoist-non-react-statics');
+
+import { parser } from './parser';
+import { OperationOption, QueryOpts, OptionProps, DataProps } from './types';
+import { default as Subscription } from './Subscriptions';
+import {
+  getDisplayName,
+  GraphQLBase,
+  calculateVariablesFromProps,
+  defaultMapPropsToOptions,
+  defaultMapPropsToSkip,
+} from './hoc-utils';
+
+export function subscribe<
+  TProps extends TGraphQLVariables | {} = {},
+  TData = {},
+  TGraphQLVariables = {},
+  TChildProps = Partial<DataProps<TData, TGraphQLVariables>>
+>(
+  document: DocumentNode,
+  operationOptions: OperationOption<TProps, TData, TGraphQLVariables, TChildProps> = {},
+) {
+  // this is memozied so if coming from `graphql` there is nearly no extra cost
+  const operation = parser(document);
+  // extract options
+  const {
+    options = defaultMapPropsToOptions,
+    skip = defaultMapPropsToSkip,
+    alias = 'Apollo',
+    shouldResubscribe,
+  } = operationOptions;
+
+  let mapPropsToOptions = options as (props: any) => QueryOpts;
+  if (typeof mapPropsToOptions !== 'function') mapPropsToOptions = () => options as QueryOpts;
+
+  let mapPropsToSkip = skip as (props: any) => boolean;
+  if (typeof mapPropsToSkip !== 'function') mapPropsToSkip = () => skip as any;
+
+  // allow for advanced referential equality checks
+  let lastResultProps: TChildProps | void;
+  return (
+    WrappedComponent: React.ComponentType<TChildProps & TProps>,
+  ): React.ComponentClass<TProps> => {
+    const graphQLDisplayName = `${alias}(${getDisplayName(WrappedComponent)})`;
+    class GraphQL extends GraphQLBase<TProps, TChildProps, { resubscribe: boolean }> {
+      static displayName = graphQLDisplayName;
+      static WrappedComponent = WrappedComponent;
+      constructor(props: TProps) {
+        super(props);
+        this.state = { resubscribe: false };
+      }
+      componentWillReceiveProps(nextProps: TProps) {
+        if (!shouldResubscribe) return;
+        this.setState({
+          resubscribe: shouldResubscribe(this.props, nextProps),
+        });
+      }
+
+      render() {
+        let props = this.props;
+        const shouldSkip = mapPropsToSkip(props);
+        const opts = shouldSkip ? Object.create(null) : mapPropsToOptions(props);
+
+        if (!shouldSkip && !Boolean(opts.variables || !operation.variables.length)) {
```

Comment:
This if statement is a little difficult to read with all the negations, it can be simplified to: 

```js
if (!shouldSkip && !opts.variables && operation.variables.length > 0) {}
```